internal/jsonschema/jsonschema.go

```diff
@@ -74,68 +74,56 @@ func validate(s *jsonschema.Schema, fsys fs.FS, path string) error {
 			return fmt.Errorf("unable to validate file %s: %w", path, err)
 		}
 
-		return newValidationErrorList(validationErr)
+		return newValidationError(validationErr)
 	}
 
 	return nil
 }
 
-func newValidationError(err *jsonschema.ValidationError) validationError {
-	path := "/"
-	if err.InstanceLocation != "" {
-		path = err.InstanceLocation
-	}
-	return validationError{
-		Path:    path,
-		Message: err.Message,
-	}
-}
-
-func newValidationErrorList(validationErr *jsonschema.ValidationError) validationErrorList {
-	if validationErr == nil {
-		return nil
-	}
+func newValidationError(err *jsonschema.ValidationError) error {
+	toVisit := err.Causes
+	var leaves []validationError
+	for ; len(toVisit) > 0; toVisit = toVisit[1:] {
+		ve := toVisit[0]
+		if ve == nil {
+			continue
+		}
 
-	if len(validationErr.Causes) == 0 {
-		return validationErrorList{newValidationError(validationErr)}
-	}
+		if len(ve.Causes) > 0 {
+			toVisit = append(toVisit, ve.Causes...)
+			continue
+		}
 
-	var errs validationErrorList
-	for _, err := range validationErr.Causes {
-		errs = append(errs, newValidationErrorList(err)...)
+		path := "/"
+		if ve.InstanceLocation != "" {
+			path = ve.InstanceLocation
+		}
+		leaves = append(leaves, validationError{path: path, message: ve.Message})
 	}
 
-	sort.Slice(errs, func(i, j int) bool {
-		return errs[i].Path > errs[j].Path
-	})
-
-	return errs
-}
-
-type validationError struct {
-	Path    string
-	Message string
-}
-
-func (e validationError) Error() string {
-	return fmt.Sprintf("%s: %s", e.Path, e.Message)
-}
+	// sort the leaves by path and message, group by path and take the first message for each path to produce a stable set of errors
```

Comment:
I think we should include all unique messages for each path rather than just the first. If there are multiple properties missing, for example, it would be nicer to be able to see that instead of playing whack-a-mole by adding the missing one we have reported and then getting a new error next time you compile.

Reply:
That's what I tried to do originally but, unfortunately, the error messages are not very good. You get whoppers like `additionalProperties 'key' not allowed, missing properties: 'apiVersion', missing properties: 'derivedRoles', missing properties: 'exportVariables', missing properties: 'principalPolicy', missing properties: 'resourcePolicy'`. 

I think it'd be quite rare to have multiple errors for a given path. We still report the errors for all  distinct paths so the user has a chance to fix all the errors at once in most cases anyway.

Reply:
It's pretty common in the playground to have multiple missing properties as you work on a policy.

Also, I think this example shows an issue: https://github.com/cerbos/cerbos/pull/1882/files#diff-1f26e131097220eb6929f987c1065dbae423d87e53ca85d4ac4e5bfb7e67bbc3

The message would be `missing properties: 'derivedRoles', missing properties: 'exportVariables', missing properties: 'principalPolicy', missing properties: 'resourcePolicy'` but we chop it down to `missing properties: 'derivedRoles'` which is sort of untrue - you could add any of the four types... and you definitely can get into this state easily in the playground because the autocompleter will happily add the `apiVersion` for you but doesn't then suggest any of the `oneof` options.